manager/assets/modext/widgets/source/modx.panel.sources.js

```diff
@@ -56,168 +59,313 @@ Ext.reg('modx-panel-sources',MODx.panel.Sources);
+                scope: this
+            }],
+            listeners: {
+                render: {
+                    fn: function(btn) {
+                        if (!this.userCanDelete) {
+                            btn.hide();
+                        }
+                    },
+                    scope: this
+                },
+                click: {
+                    fn: function(btn) {
+                        const removableSources = this.getRemovableItemsFromSelection('int'),
+                              menuOptRemove = btn.menu.getComponent('modx-bulk-menu-opt-remove')
+                        ;
+                        // console.log('removableSources: ',removableSources);
+                        // console.log('removableSources empty? ',Ext.isEmpty(removableSources));
+                        // console.log('menuOptRemove: ',menuOptRemove);
+                        if (removableSources.length === 0) {
+                            menuOptRemove.disable();
+                        } else {
+                            menuOptRemove.enable();
+                        }
+                    },
+                    scope: this
+                }
+            }
+        }, 
+        '->', 
+        this.getQueryFilterField(),
+        this.getClearFiltersButton()
+        ],
+        viewConfig: {
+            forceFit: true,
+            scrollOffset: 0,
+            getRowClass: function(record, index, rowParams, store) {
+                // Adds the returned class to the row container's css classes
+                if (this.grid.userCanDeleteRecord(record)) {
+                    return '';
+                }
+                const rowClasses = 'disable-selection';
+                return record.json.isProtected ? `modx-protected-row  ${rowClasses}` : rowClasses ;
+            }
+        }
+    });
+    MODx.grid.Sources.superclass.constructor.call(this, config);
+
+    this.protectedDataIndex = 'id';
+    this.protectedIdentifiers = [1];
+    this.gridMenuActions = ['edit', 'delete', 'duplicate'];
+
+    this.setUserCanEdit(['source_save', 'source_edit']);
+    this.setUserCanCreate(['source_save']);
+    this.setUserCanDelete(['source_delete']);
+    this.setShowActionsMenu();
+
+    this.on({
+        render: function(grid) {
+            this.setEditableColumnAccess(
+                ['modx-source--name', 'modx-source--description']
+            );
+        },
+        beforeedit: function(e){
+            // if (e.record.json.name === 'Filesystem' || !this.userCanEditRecord(e.record)) {
+            if (e.record.json.isProtected || !this.userCanEditRecord(e.record)) {
+                return false;
+            }
+        }
+    });
+
+    this.getStore().on({
+        load: function(store, records, params){
+            records.forEach(record => {
+                if (!this.userCanDeleteRecord(record)) {
+                    this.nonRemoveableRecords.push(record.id);
```

Comment:
@smg6511 this line causes an error as `nonRemovableRecords` is undefined

Reply:
Ok, this is fixed. I'd actually gone in a little different direction since submitting the original PR, and this listener is no longer used.